demos/faults.cc

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under both the 3-Clause BSD License and the GPLv2, found in the
+ * LICENSE and LICENSE.GPL-2.0 files, respectively, in the root directory.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0
+ */
+
+#include "faults.h"
+
+#include <setjmp.h>
+
+#include <cstring>
+
+namespace {
+
+// The fault handler that should be called from the signal handler, and the
```

Comment:
Unfortunately, making them `thread_local` makes them UB to access within a signal handler, per [[support.signal]](https://eel.is/c++draft/support.signal) (3.2). Since it was only ever aspirational anyway, just delete.

Reply:
So, a few things about this and other comments on signal safety.

First, yes, getting async signal safety right, to the standards of the sort of systems where it usually comes up, is very hard. That puts us in tension because the more we work to make this "right", the harder it is to understand.

For our application, here, I think the right tradeoff is to be very easy to understand, and at least _mostly_ right. Obviously we'll fix bugs or unreliability we hit on the platforms we support.

There are some other mitigating factors for us, aside from the obvious ones like "our entire repository is predicated on the existence of undefined but predictable behavior". The signals we care about, faults like `SIGSEGV`, are _synchronous_ -- they are caused by instructions in the program, rather than external events.

These signals are also usually _fatal_, so we can say with some confidence that they are not normally raised by code not under our control, like the standard library. They also will not be raised during execution of the signal handler itself unless we have a bug, in which case we get the failure we wanted.

All this to say: for now, however distasteful it might seem, I suggest we use "works as expected in concrete implementations" as our standard rather than "works on the abstract machine". I would be fine adding comments to the code to that effect.

Fun trivia - Rust also accesses thread-local data in a `SIGSEGV` handler. https://github.com/rust-lang/rust/issues/43146

Reply:
I still removed `thread_local`.

Reply:
FWIW I think atomics would be likely to work with our use case, without any undefined, unspecified, or implementation-defined behavior.

There's one other issue that made me cautious that I should've mentioned first: different platforms send signals to different threads, so the moment we have thread-local signal handlers, even if there's no signals raised during assignment to the threadlocal we're in Very Difficult territory, which makes it harder, rather than easier, to understand.